pmk/scripts/migrate_awsvolumes/migrate_ebs_volume_gp2_to_gp3.py

```diff
@@ -0,0 +1,297 @@
+
+    def migrate_volume_to_GP3(self, instance_id):
+        logger.info(
+            "Batch migrate_volume_to_GP3 for instance: %s ", instance_id)
+        filter = [{
+            'Name': 'attachment.instance-id',
+            'Values': [instance_id, ]
+        }]
+
+        response = self.ec2_client.describe_volumes(Filters=filter)
+        for vol in response["Volumes"]:
+            volume_id = vol["VolumeId"]
+            if vol["VolumeType"] == "gp2":
+                logger.info("Modifiying volume: %s", volume_id)
+                self.ec2_client.modify_volume(
+                    VolumeId=volume_id, VolumeType='gp3')
+            self.check_modification_status(volume_id)
+
+    def batch_migrate_volumes(self, instance_ids):
+        threads = []
+        for i in range(len(instance_ids)):
+            thread = threading.Thread(target=self.migrate_volume_to_GP3,
+                                      args=(instance_ids[i],))
+            threads.append(thread)
+
+        for i in threads:
+            i.start()
+
+        for i in threads:
+            i.join()
+
+
+if __name__ == '__main__':
+
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--kdu', type=str, required=True)
+    parser.add_argument('--tenant', type=str, required=True)
+    parser.add_argument('--cluster_id', type=str, required=True)
+    parser.add_argument('--username', default="", type=str, required=True)
+    parser.add_argument('--password', default="", type=str, required=True)
+
+    args = parser.parse_args()
+
+    qbertClient = QbertAPI(args.kdu, args.tenant, args.username, args.password)
+    cluster_id = args.cluster_id
+    migratevol = MigrateVolume()
+    batchSize = 5
+
+    try:
+        # get cluster
+        logger.info("Fetching cluster [%s] information.", cluster_id)
+        cluster_resp = qbertClient.get_cluster_by_uuid(cluster_id)
+        logger.info("cluster %s", cluster_resp.json()["name"])
+
+        master_nodes, worker_nodes = qbertClient.get_nodes_by_cluster_uuid(
+            cluster_id)
+
+        if not (master_nodes and worker_nodes):
+            logger.warning(
+                "cluster %s doesn't have any master/worker nodes to migrate volume", cluster_resp.json()["name"])
+
+        logger.info("Master nodes %s Worker nodes %s ",
+                    master_nodes, worker_nodes)
+        # Migrate volumes for master nodes
+        for master_node in master_nodes:
+            logger.info("Migrating masternode (%s) volumes to GP3",
+                        cluster_resp.json()["name"])
+            migratevol.migrate_volume_to_GP3(master_node)
+
+        # Migrate volumes for worker nodes in batch of 5
+        for i in range(0, len(worker_nodes), batchSize):
+            logger.info("Migrating workernodes (%s) to GP3 GP3 volumetype",
+                        worker_nodes[i:i + batchSize])
+            migratevol.batch_migrate_volumes(worker_nodes[i:i + batchSize])
```

Comment:
Since the batch operation runs in a thread, will the caller here be able to catch those exceptions in the child thread?
From my understanding exceptions handling happens only for own context/stack.

Reply:
@psarwate  if any thread raises an exception that will log an error/traceback and the rest of the thread will continue to work!

Reply:
Added try-except block!